lib/proj/do_proj.c

```diff
@@ -83,95 +91,403 @@ int GPJ_init_transform(const struct pj_info *info_in,
-	    char *insrid, *outsrid;
+    info_trans->meters = 1.;
+    info_trans->zone = 0;
+    sprintf(info_trans->proj, "pipeline");
 
-#if PROJ_VERSION_MAJOR >= 6
-	    /* PROJ6+: EPSG must uppercase EPSG */
-	    if (strncmp(info_in->srid, "epsg", 4) == 0)
-		insrid = G_store_upper(info_in->srid);
-	    else
-		insrid = G_store(info_in->srid);
+    /* PROJ6+: EPSG must uppercase EPSG */
+    if (info_in->srid) {
+	if (strncmp(info_in->srid, "epsg", 4) == 0)
+	    insrid = G_store_upper(info_in->srid);
+	else
+	    insrid = G_store(info_in->srid);
+    }
 
-	    if (strncmp(info_out->srid, "epsg", 4) == 0)
-		outsrid = G_store_upper(info_out->srid);
-	    else
-		outsrid = G_store(info_out->srid);
+    if (info_out->pj && info_out->srid) {
+	if (strncmp(info_out->srid, "epsg", 4) == 0)
+	    outsrid = G_store_upper(info_out->srid);
+	else
+	    outsrid = G_store(info_out->srid);
+    }
 
-	    /* PROJ6+: enforce axis order easting, northing
-	     * +axis=enu (works with proj-4.8+) */
+    /* user-provided pipeline */
+    if (info_trans->def) {
+	/* create a pj from user-defined transformation pipeline */
+	info_trans->pj = proj_create(PJ_DEFAULT_CTX, info_trans->def);
+	if (info_trans->pj == NULL) {
+	    G_warning(_("proj_create() failed for '%s'"), info_trans->def);
 
-#else
-	    /* PROJ5: EPSG must lowercase epsg */
-	    if (strncmp(info_in->srid, "EPSG", 4) == 0)
-		insrid = G_store_lower(info_in->srid);
-	    else
-		insrid = G_store(info_in->srid);
+	    return -1;
+	}
+	projstr = proj_as_proj_string(NULL, info_trans->pj, PJ_PROJ_5, NULL);
+	if (projstr == NULL) {
+	    G_warning(_("proj_create() failed for '%s'"), info_trans->def);
 
-	    if (strncmp(info_out->srid, "EPSG", 4) == 0)
-		outsrid = G_store_lower(info_out->srid);
-	    else
-		outsrid = G_store(info_out->srid);
+	    return -1;
+	}
+	else {
+	    /* make sure axis order is easting, northing 
+	     * proj_normalize_for_visualization() does not work here 
+	     * because source and target CRS are unknown to PROJ
+	     * remove any "+step +proj=axisswap +order=2,1" ?
+	     *  */
+	    info_trans->def = G_store(projstr);
+
+	    if (strstr(info_trans->def, "axisswap")) {
+		G_warning(_("The transformation pipeline contains an '%s' step. "
+			    "Remove this step if easting and northing are swapped in the output."),
+			    "axisswap");
+	    }
+
+	    G_debug(1, "proj_create() pipeline: %s", info_trans->def);
+
+	    /* the user-provided PROJ pipeline is supposed to do 
+	     * all the needed unit conversions */
+	    /* ugly hack */
+	    ((struct pj_info *)info_in)->meters = 1;
+	    ((struct pj_info *)info_out)->meters = 1;
+	}
+    }
+    /* if no output CRS is defined, 
+     * assume info_out to be ll equivalent of info_in */
+    else if (info_out->pj == NULL) {
+	/* what about axis order?
+	 * is it always enu? 
+	 * probably yes, as long as there is no +proj=axisswap step */
+	G_asprintf(&(info_trans->def), "+proj=pipeline +step +inv %s",
```

Comment:
in the case where info_in->def would contain a +towgs84/+nadgrids clause, I believe this pipeline would apply it, whereas you probably only want the reverse projection, and no datum shift.
The easiest would probably to mess up with the PROJ string. Otherwise with the PROJ API, you could instanciate a PJ object from the string, check if it is a BoundCRS with proj_get_source_crs(), and in that case, take the source CRS with proj_get_source_crs(), and do the inverse transform on it